lib/register.php

```diff
@@ -117,3 +191,77 @@ function gutenberg_add_edit_links( $actions, $post ) {
 
 	return $actions;
 }
+
+/**
+ * Get the edit post URL for Gutenberg.
+ *
+ * @since 0.5.0
+ *
+ * @param int $post_id Post ID.
+ * @return string|false URL or false if not available.
+ */
+function gutenberg_get_edit_post_url( $post_id ) {
+	// Note that menu_page_url() cannot be used because it does not work on the frontend.
+	$gutenberg_url = admin_url( 'admin.php?page=gutenberg' );
+	$gutenberg_url = add_query_arg( 'post_id', $post_id, $gutenberg_url );
+	return $gutenberg_url;
+}
+
+/**
+ * Filters the post edit link to default to the Gutenberg editor when the post content contains a block.
+ *
+ * @since 0.5.0
+ *
+ * @param string $url     The edit link URL.
+ * @param int    $post_id Post ID.
+ * @param string $context The link context. If set to 'display' then ampersands are encoded.
+ * @return string Edit post link.
+ */
+function gutenberg_filter_edit_post_link( $url, $post_id, $context ) {
+	$post = get_post( $post_id );
+	if ( gutenberg_can_edit_post( $post_id ) && gutenberg_post_has_blocks( $post_id ) ) {
+		$gutenberg_url = gutenberg_get_edit_post_url( $post->ID );
+		if ( 'display' === $context ) {
+			$gutenberg_url = esc_url( $gutenberg_url );
+		}
+		$url = $gutenberg_url;
+	}
+	return $url;
+}
+add_filter( 'get_edit_post_link', 'gutenberg_filter_edit_post_link', 10, 3 );
+
+/**
+ * Return whether the post can be edited in Gutenberg and by the current user.
+ *
+ * Gutenberg depends on the REST API, and if the post type is not shown in the
+ * REST API, then the post cannot be edited in Gutenberg.
+ *
+ * @since 0.5.0
+ *
+ * @param int $post_id Post.
+ * @return bool Whether the post can be edited with Gutenberg.
+ */
+function gutenberg_can_edit_post( $post_id ) {
+	$post = get_post( $post_id );
+	if ( ! $post || ! post_type_exists( $post->post_type ) ) {
+		return false;
+	}
+	$post_type_object = get_post_type_object( $post->post_type );
+	if ( ! $post_type_object->show_in_rest ) {
+		return false;
+	}
+	return current_user_can( 'edit_post', $post_id );
+}
+
+/**
+ * Determine whether a post has blocks.
+ *
+ * @since 0.5.0
+ *
+ * @param int $post_id Post ID.
+ * @return bool Whether the post has blocks.
+ */
+function gutenberg_post_has_blocks( $post_id ) {
+	$post = get_post( $post_id );
+	return $post && preg_match( '#<!-- wp:\w+#', $post->post_content );
```

Comment:
Yeah, I think that if `<!-- wp:` is already there, the chance is very high that it is followed by a word-like sequence anyway.